packages/kbn-search-api-keys-components/src/providers/search_api_key_provider.tsx

```diff
@@ -0,0 +1,182 @@
+  switch (action.type) {
+    case 'SET_API_KEY':
+      return { ...state, apiKey: action.apiKey, status: action.status };
+    case 'SET_STATUS':
+      return { ...state, status: action.status };
+    case 'TOGGLE_API_KEY_VISIBILITY':
+      return {
+        ...state,
+        status:
+          state.status === Status.showHiddenKey ? Status.showPreviewKey : Status.showHiddenKey,
+      };
+    case 'CLEAR_API_KEY':
+      return { ...state, apiKey: null, status: Status.showCreateButton };
+    default:
+      return state;
+  }
+};
+
+export const ApiKeyContext = createContext<APIKeyContext>({
+  displayedApiKey: null,
+  apiKey: null,
+  toggleApiKeyVisibility: () => {},
+  updateApiKey: () => {},
+  status: Status.uninitialized,
+  apiKeyIsVisible: false,
+  initialiseKey: () => {},
+});
+
+export const SearchApiKeyProvider: React.FC = ({ children }) => {
+  const { http } = useKibana().services;
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  const updateApiKey = useCallback(({ id, encoded }: { id: string; encoded: string }) => {
+    sessionStorage.setItem(API_KEY_STORAGE_KEY, JSON.stringify({ id, encoded }));
+    dispatch({ type: 'SET_API_KEY', apiKey: encoded, status: Status.showHiddenKey });
+  }, []);
+  const handleShowKeyVisibility = useCallback(() => {
+    dispatch({ type: 'TOGGLE_API_KEY_VISIBILITY' });
+  }, []);
+  const initialiseKey = useCallback(() => {
+    dispatch({ type: 'SET_STATUS', status: Status.loading });
+  }, []);
+  const { mutateAsync: validateApiKey } = useMutation(async (id: string) => {
+    try {
+      if (!http?.post) {
+        throw new Error('HTTP service is unavailable');
+      }
+
+      const response = await http.post<{ isValid: boolean }>(APIRoutes.API_KEY_VALIDITY, {
+        body: JSON.stringify({ id }),
+      });
+
+      return response.isValid;
+    } catch (err) {
+      return false;
+    }
+  });
+  const { mutateAsync: createApiKey } = useMutation<APIKeyCreationResponse | undefined>({
+    mutationFn: async () => {
+      try {
+        if (!http?.post) {
+          throw new Error('HTTP service is unavailable');
+        }
+
+        return await http.post<APIKeyCreationResponse>(APIRoutes.API_KEYS);
+      } catch (err) {
+        if (err.response?.status === 400) {
+          dispatch({ type: 'SET_STATUS', status: Status.showCreateButton });
+        } else if (err.response?.status === 403) {
+          dispatch({ type: 'SET_STATUS', status: Status.showUserPrivilegesError });
+        } else {
+          throw err;
+        }
+      }
+    },
+    onSuccess: (receivedApiKey) => {
+      if (receivedApiKey) {
+        sessionStorage.setItem(
+          API_KEY_STORAGE_KEY,
+          JSON.stringify({ id: receivedApiKey.id, encoded: receivedApiKey.encoded })
+        );
+        dispatch({
+          type: 'SET_API_KEY',
+          apiKey: receivedApiKey.encoded,
+          status: Status.showHiddenKey,
+        });
+      }
+    },
+  });
+
+  useEffect(() => {
```

Comment:
since we're rendering this provider in the `search_indices` application now this will be called for all search_indices routes on page load.

With this going out to serverless only for now, I'm ok with it for now. But I think we'll need to refine this in the future to be more explicit about when the creation logic is called vs when we just check and read for an existing key.

Reply:
yep this useEffect will run on page routes but the API creation code will only be executed once a hook has been mounted and will only be run once until the provider is unmounted.